File: library/trunk/src/org/webpki/json/JSONArrayReader.java
/*
 *  Copyright 2006-2013 WebPKI.org (http://webpki.org).
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
package org.webpki.json;

import java.io.IOException;
import java.math.BigDecimal;
import java.math.BigInteger;

import java.util.Vector;

/**
 * For writing array elements.
 */
public class JSONArrayReader
  {
    Vector<JSONValue> array;

    int index;

    JSONArrayReader (Vector<JSONValue> array)
      {
        this.array = array;
      }

    public boolean hasMore ()
      {
        return index < array.size ();
      }

    void inRangeCheck () throws IOException
      {
        if (!hasMore ())
          {
            throw new IOException ("Trying to read past of array limit: " + index);
          }
      }
    
    Object get (JSONTypes expected_type) throws IOException
      {
        inRangeCheck ();
        JSONValue value = array.elementAt (index++);
        if (!expected_type.isCompatible (value.type))
          {
            throw new IOException ("Incompatible request: " + expected_type + " versus " + value.type);
          }
        return value.value;
      }

    public String getString () throws IOException
      {
        return (String) get (JSONTypes.STRING);
      }

    public int getInt () throws IOException
      {
        return Integer.parseInt ((String) get (JSONTypes.INTEGER));
      }

    public BigInteger getBigInteger () throws IOException
      {
        return new BigInteger ((String) get (JSONTypes.INTEGER));
      }

    public BigDecimal getBigDecimal () throws IOException
      {
        return new BigDecimal ((String) get (JSONTypes.DECIMAL));
      }

    public double getDouble () throws IOException
      {
        return new Double ((String) get (JSONTypes.FLOATING_POINT));
      }

    public boolean getBoolean () throws IOException
      {
        return new Boolean ((String) get (JSONTypes.BOOLEAN));
      }

    public boolean getIfNULL () throws IOException
      {
        if (getElementType () == JSONTypes.NULL)
          {
            scanAway ();
            return true;
          }
        return false;
      }

    @SuppressWarnings("unchecked")
    public JSONArrayReader getArray () throws IOException
      {
        return new JSONArrayReader ((Vector<JSONValue>)get (JSONTypes.ARRAY));
      }

    public JSONTypes getElementType () throws IOException
      {
        inRangeCheck ();
        return array.elementAt (index).type;
      }

    public JSONObjectReader getObject () throws IOException
      {
        return new JSONObjectReader ((JSONObject) get (JSONTypes.OBJECT));
      }

    public void scanAway () throws IOException
      {
        get (getElementType ());
      }
  }

File: javascript/trunk/src/jsonparser/JSONArrayReader.js
/*
 *  Copyright 2006-2014 WebPKI.org (http://webpki.org).
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */

/*================================================================*/
/*                        JSONArrayReader                         */
/*================================================================*/

org.webpki.json.JSONArrayReader = function (/* Vector<org.webpki.json.JSONValue> */array)
{
    this.index = 0;
    this.array = array;
};

/* public boolean */org.webpki.json.JSONArrayReader.prototype.hasMore = function ()
{
    return this.index < this.array.length;
};

/* void */org.webpki.json.JSONArrayReader.prototype._inRangeCheck = function ()
{
    if (!this.hasMore ())
    {
        org.webpki.json.JSONError._error ("Trying to read past of array limit: " + this.index);
    }
};

/* Object */org.webpki.json.JSONArrayReader.prototype._get = function (/* org.webpki.json.JSONTypes */expected_type)
{
    this._inRangeCheck ();
    /* org.webpki.json.JSONValue */var value = this.array[this.index++];
    if (!expected_type.isCompatible (value.type))
    {
        org.webpki.json.JSONError._error ("Incompatible request: " +
                "Read=" + org.webpki.json.JSONValue.prototype.getJSONTypeName (value.type) +
                ", Expected=" + org.webpki.json.JSONValue.prototype.getJSONTypeName (expected_type));
    }
    return value.value;
};

/* public String */org.webpki.json.JSONArrayReader.prototype.getString = function ()
{
    return /* (String) */this._get (org.webpki.json.JSONTypes.STRING);
};

/* public int */org.webpki.json.JSONArrayReader.prototype.getInt = function ()
{
    return parseInt (/* (String) */this._get (org.webpki.json.JSONTypes.INTEGER));
};

/* BigInteger */org.webpki.json.JSONArrayReader.prototype.getLong = function ()
{
    return this.getBigInteger ().getLong ();
};

/* public BigInteger */org.webpki.json.JSONArrayReader.prototype.getBigInteger = function ()
{
    return org.webpki.math.BigInteger.fromString (this._get (org.webpki.json.JSONTypes.INTEGER));
};

//No real support for BigDecimal but at least text parsing is performed

/* public BigDecimal */org.webpki.json.JSONArrayReader.prototype.getBigDecimal = function ()
{
    return this._get (org.webpki.json.JSONTypes.DECIMAL);
};

/* public Date */org.webpki.json.JSONArrayReader.prototype.getDateTime = function ()
 {
    return new Date (this.getString ());
 };

 /* public double */org.webpki.json.JSONArrayReader.prototype.getDouble = function ()
 {
     return parseFloat (this._get (org.webpki.json.JSONTypes.DOUBLE));
 };

 /* public boolean */org.webpki.json.JSONArrayReader.prototype.getBoolean = function ()
 {
     return this._get (org.webpki.json.JSONTypes.BOOLEAN) == "true";
 };

 /* public boolean */org.webpki.json.JSONArrayReader.prototype.getIfNULL = function ()
 {
     if (this.getElementType () == org.webpki.json.JSONTypes.NULL)
     {
         this.scanAway ();
         return true;
     }
     return false;
 };
 
/* public org.webpki.json.JSONArrayReader */org.webpki.json.JSONArrayReader.prototype.getArray = function ()
{
    return new org.webpki.json.JSONArrayReader (/* (Vector<org.webpki.json.JSONValue>) */this._get (org.webpki.json.JSONTypes.ARRAY));
};

/* public org.webpki.json.JSONTypes */org.webpki.json.JSONArrayReader.prototype.getElementType = function ()
{
    this._inRangeCheck ();
    return this.array[this.index].type;
};

/* public org.webpki.json.JSONObjectReader */org.webpki.json.JSONArrayReader.prototype.getObject = function ()
{
    return new org.webpki.json.JSONObjectReader (/* (org.webpki.json.JSONObject) */this._get (org.webpki.json.JSONTypes.OBJECT));
};

/* public void */org.webpki.json.JSONArrayReader.scanAway = function ()
{
    this._get (this.getElementType ());
};
